src/IO/ImageDataFromChunks.worker.js

```diff
@@ -46,74 +42,82 @@ registerWebworker().operation(
-      ]
-      const itChunkOffsets = [0, 0, 0, 0]
-      itChunkOffsets[3] = chunkStrides[3] * l
-      const itPixelOffsets = [0, 0, 0]
-      for (let kk = itStart[2]; kk < itEnd[2]; kk++) {
-        itChunkOffsets[2] = chunkStrides[2] * (kk - k * chunkSize[3])
-        itPixelOffsets[2] = pixelStrides[2] * (kk - indexStart[2])
-        for (let jj = itStart[1]; jj < itEnd[1]; jj++) {
-          itChunkOffsets[1] = chunkStrides[1] * (jj - j * chunkSize[2])
-          itPixelOffsets[1] = pixelStrides[1] * (jj - indexStart[1])
-          for (let ii = itStart[0]; ii < itEnd[0]; ii++) {
-            const begin =
-              chunkStrides[0] * (itStart[0] - i * chunkSize[1]) +
-              itChunkOffsets[1] +
-              itChunkOffsets[2] +
-              itChunkOffsets[3]
-            const end = begin + components * (itEnd[0] - itStart[0])
-            const offset =
-              pixelStrides[0] * (itStart[0] - indexStart[0]) +
-              itPixelOffsets[1] +
-              itPixelOffsets[2]
-            const subarray = chunk.subarray(begin, end)
-            pixelArray.set(subarray, offset)
-          } // for every column
-        } // for every row
-      } // for every slice
-    }
+      const chunkEnd = {
+        c: (c + 1) * chunkSize.c,
+        z: (z + 1) * chunkSize.z,
+        y: (y + 1) * chunkSize.y,
+        x: (x + 1) * chunkSize.x,
+      }
+      const roiStart = Object.fromEntries(indexStart)
+      const roiEnd = Object.fromEntries(indexEnd)
+      validateIndices({ chunkStart, chunkEnd, roiStart, roiEnd })
+
+      // iterate on image from chunk or ROI start
+      const itStart = {
+        c: Math.max(chunkStart.c, roiStart.c),
+        z: Math.max(chunkStart.z, roiStart.z),
+        y: Math.max(chunkStart.y, roiStart.y),
+        x: Math.max(chunkStart.x, roiStart.x),
+      }
+      const itEnd = {
+        c: Math.min(chunkEnd.c, roiEnd.c),
+        z: Math.min(chunkEnd.z, roiEnd.z),
+        y: Math.min(chunkEnd.y, roiEnd.y),
+        x: Math.min(chunkEnd.x, roiEnd.x),
+      }
+
+      for (let cc = itStart.c; cc < itEnd.c; cc++) {
+        for (let zz = itStart.z; zz < itEnd.z; zz++) {
+          for (let yy = itStart.y; yy < itEnd.y; yy++) {
+            for (let xx = itStart.x; xx < itEnd.x; xx++) {
+              pixelArray[
+                cc +
+                  xx * pixelStrides.x +
+                  yy * pixelStrides.y +
+                  zz * pixelStrides.z
+              ] =
+                chunk[
+                  // subtract x * chunkSize.x from xx to start at beginning of chunk despite itStart.x
+                  (xx - x * chunkSize.x) * chunkStrides.x +
+                    (yy - y * chunkSize.y) * chunkStrides.y +
+                    (zz - z * chunkSize.z) * chunkStrides.z +
+                    (cc - c * chunkSize.c) * chunkStrides.c
+                ]
+            } // column
+          } // row
+        } // slice
+      } // component
+    } // chunk
```

Comment:
Optimizations to add to this critical piece:

[ ] 1. Three versions of the loop based on the dim order. a) tczyx, b) tzyxc c) other. a) for most OME-NGFF, b for the cancel contiguous, we may encounter c) in the wild, e.g. it is also common to have tzcyz. c) can be like it is now. a) and b) will further be optimized in the following steps.

[ ] 2. For the conditions where we can use them in the inner-most loop, use `.subarray` and `.set` for getting and setting the contiguous region.

[ ] 3. Cache the offsets in the intermediate loops.

Reply:
Added caching of offsets.  Do we have any example images of non `tczyx` dim ordering we would like to support?  I have not tested any...